### PR TITLE
Add overview subsection

### PR DESCRIPTION
**Pull request by gramalingam**
Add a short overview subsection to discuss and cover semantic and operational aspects of the API. (This is primarily motivated by the discussion in the issue about introducing terminology, such as views: https://github.com/webmachinelearning/webnn/issues/150 )

***
<a href="https://pr-preview.s3.amazonaws.com/gramalingam/webnn/pull/196.html" title="Last updated on Aug 3, 2021, 6:19 PM UTC (0471426)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/196/31a71f6...gramalingam:0471426.html" title="Last updated on Aug 3, 2021, 6:19 PM UTC (0471426)">Diff</a>

### CODE REVIEW

**Review by huningxin — APPROVED** [review]
Thanks @gramalingam for this significant contribution. The content looks good.

There are bikeshed build errors. I fixed them and submitted a [PR](https://github.com/gramalingam/webnn/pull/1/files) to your branch. Please take a look.

File: index.bs
Comment (wchao1115) [inline index.bs:409]: Thanks @gramalingam. The notion of `MLOperand` is a bit harder to explain. But I think it helps to draw a contrast between what is an operand and what is a tensor. In my mind, an operand is really a *description* of the data and not the data itself i.e. we need to bind the data to operands before executing the model, etc. I sense that's probably what you try to say in your wordings around "is a static (compile-time) representation"?

File: index.bs
Comment (wchao1115) [inline index.bs:425]: Similar to the feedback above, I think what missing here is the bind step. For example we can explain it as "before the compiled graph is executed, the actual tensors, which are multidimensional arrays of binary data, need to be bound to the operands in the graph...", etc.

File: index.bs
Comment (gramalingam) [reply index.bs:410]: Thanks, addressed. Apologies for the delay in responding, something is wrong with my notification/email, I didn't see the notifications for these comments.

File: index.bs
Comment (gramalingam) [reply index.bs:433]: Addressed, thanks

**Comment by gramalingam** [comment]
@huningxin @wchao1115 : please let me know if you have any further comments, thanks!

File: index.bs
Comment (wchao1115) [inline index.bs:410]: Thanks for updating the wordings. I just think that it's a lot simpler to explain `MLOperand` as a *description* of the data to be processed by the graph at execution time, and not the data itself aka the *tensors*. Hopefully that saves you from having to invoke the *in programming language terminology..* paradigm, which is rather conceptual in my view.

Comment (gramalingam) [reply index.bs:410]: Thanks Chai for the suggestion. The problem I have is drawing yet another distinction: A _tensor_ itself  arguably consists of _data_ and a description of the data (such as shape, for example). So, I find the terms "data" and "description of data" themselves to be ambiguous (with these two different interpretations). I couldn't find a simple way to summarize this unambiguously. The key distinguishing feature of an MLOperand is that this is a description that is available prior to execution-time (that is, a static or compile-time description). I realize that I used the word "represention" above. Would it help to change it to ```static (compile-time) description of runtime data```?

Comment (huningxin) [reply index.bs:410]: @wchao1115 
> `MLOperand` as a _description_ of the data to be processed by the graph at execution time, and not the data itself aka the _tensors_

This is not always true. For [constant `MLOperand`](https://webmachinelearning.github.io/webnn/#dom-mlgraphbuilder-constant), it contains both description and data.

@gramalingam
> I realize that I used the word "representation" above.

I think "presentation" is good. To me, the computation graph contains the tensor operations. The `MLOperand` just represents the tensor of operation when building the computational graph. They could represent input, output, constant and intermediate tensor. At compute-time, the caller should bind the value to the input `MLOperand` and provide the pre-allocated buffer to the output `MLOperand` that receives the value.

File: index.bs
Comment (huningxin) [inline index.bs:411]: `MLOperand` could also represent constant data that is not a variable.

File: index.bs
Comment (huningxin) [inline index.bs:428]: The caller would also provide the pre-allocated buffers to output `MLOperand` by using `MLNamedOutputs`.

File: index.bs
Comment (wchao1115) [reply index.bs:410]: > For constant MLOperand, it contains both description and data.

True.

> I couldn't find a simple way to summarize this unambiguously.

What if we just focus on explaining the nature of an operand as a representation of the data that flows within the computational graph including the constant, input, output, and intermediate data between two operations? I feel that by describing it directly this way, the reference to the programming language terminology could be set aside.

Comment (huningxin) [reply index.bs:410]: > What if we just focus on explaining the nature of an operand as a representation of the data that flows within the computational graph including the constant, input, output, and intermediate data between two operations?

sgtm

Comment (gramalingam) [reply index.bs]: Sounds good to me too. I updated it accordingly. Please take a look. Thanks!

File: index.bs
Comment (gramalingam) [reply index.bs:426]: Added this, thanks. There is one thing I don't understand: I assume that the ```sequence<long> dimensions``` in ```MLInput``` represents shape information. It looks like it is optional for inputs and missing for outputs. Can you clarify the intention? Thanks!

**Review by wchao1115 — APPROVED** [review]

**Comment by gramalingam** [comment]
Hi @wchao1115  @huningxin : it says the workflow is waiting for approval from a maintainer: can you please take a look? Thanks!

**Review by huningxin — APPROVED** [review]

File: index.bs
Comment (huningxin) [reply index.bs:426]: The `dimensions` of `MLInput` is useful for the dynamic input shape use case. The call can specify the concrete input shape when compute. For output, it expects the caller to provide the pre-allocated output buffer, it assumes the caller (usually the framework) knows the output shape and allocates the output buffer.

Do you see anything missing?

**Comment by huningxin** [comment]
> Hi @wchao1115 @huningxin : it says the workflow is waiting for approval from a maintainer: can you please take a look? Thanks!

@gramalingam , thanks for addressing the feedbacks. The PR has been approved and merged.

File: index.bs
Comment (gramalingam) [reply index.bs:426]: In all DNN frameworks, the Tensor abstraction combines a pointer to some memory (the data or the buffer), with a minimum of shape (the dimensions). Aside from memory-allocation, most tensor-processing is dependent on the shape information. So, it seems to me that such an abstraction would be useful in this specification too.

Comment (huningxin) [reply index.bs:426]: > Aside from memory-allocation, most tensor-processing is dependent on the shape information.

I agree. The shape information is provided in the graph build time. At compute time, only buffer needs to be bound except for the input with dynamic shape, it needs both data and concrete shape.

**Comment by anssiko** [comment]
@gramalingam thank you for this substantial contribution! This PR (including the discussion in this PR) is important documentation. See also https://github.com/webmachinelearning/webnn/pull/201 and propose adjustments to the explainer text as needed.

@huningxin @wchao1115 please add @gramalingam's name to https://webmachinelearning.github.io/webnn/#acknowledgements in recognition of this substantial contribution and also think of other contributors that may be missing. I'd encourage us to actively recognize contributors in the Acknowledgements section. This is an effort that builds on work of many.

**Comment by anssiko** [comment]
Good catch :-)

Let my comment serve as a general reminder the editors have a mandate to update the Acknowledgements section in any PR that incorporates helpful contributions from other group participants.

**Pull request merged**
